src/storage/s3.rs

```diff
@@ -18,51 +19,60 @@ mod test;
 pub(crate) use test::TestS3;
 
 pub(crate) static S3_BUCKET_NAME: &str = "rust-docs-rs";
-static S3_RUNTIME: Lazy<Mutex<Runtime>> =
-    Lazy::new(|| Mutex::new(Runtime::new().expect("Failed to create S3 runtime")));
+static S3_RUNTIME: Lazy<Runtime> =
+    Lazy::new(|| Runtime::new().expect("Failed to create S3 runtime"));
 
 pub(crate) struct S3Backend {
-    client: S3Client,
+    pub client: S3Client,
     bucket: String,
+    pub runtime: &'static Runtime,
 }
 
 impl S3Backend {
     pub(crate) fn new(client: S3Client, bucket: &str) -> Self {
         Self {
             client,
             bucket: bucket.into(),
+            runtime: &*S3_RUNTIME,
         }
     }
 
     pub(super) fn get(&self, path: &str, max_size: usize) -> Result<Blob, Error> {
-        let res = self
-            .client
-            .get_object(GetObjectRequest {
-                bucket: self.bucket.to_string(),
-                key: path.into(),
-                ..Default::default()
-            })
-            .sync()?;
+        self.runtime.handle().block_on(async {
+            let res = self
+                .client
+                .get_object(GetObjectRequest {
+                    bucket: self.bucket.to_string(),
+                    key: path.into(),
+                    ..Default::default()
+                })
+                .await?;
 
-        let mut content = crate::utils::sized_buffer::SizedBuffer::new(max_size);
-        content.reserve(
-            res.content_length
-                .and_then(|l| l.try_into().ok())
-                .unwrap_or(0),
-        );
+            let mut content = crate::utils::sized_buffer::SizedBuffer::new(max_size);
+            content.reserve(
+                res.content_length
+                    .and_then(|l| l.try_into().ok())
+                    .unwrap_or(0),
+            );
+
+            let mut body = res
+                .body
+                .ok_or_else(|| failure::err_msg("Received a response from S3 with no body"))?;
 
-        let mut body = res.body.unwrap().into_blocking_read();
-        std::io::copy(&mut body, &mut content)?;
+            while let Some(data) = body.next().await.transpose()? {
+                content.write_all(data.as_ref())?;
+            }
```

Comment:
This calls `await` on each chunk in serial, which seems needlessly slow. Can we try to `await` all of them at once somehow? That's what `into_blocking_read` did before.

As an aside, where did you find `next()`? I don't see it on https://docs.rs/rusoto_core/0.45.0/rusoto_core/struct.ByteStream.html.

Reply:
`.next()` comes from the [`Stream` impl](https://docs.rs/rusoto_core/0.45.0/rusoto_core/struct.ByteStream.html#impl-Stream). This is also doing the [same thing `.into_blocking_read` does](https://docs.rs/rusoto_signature/0.45.0/src/rusoto_signature/stream.rs.html#139-141). With maintaining the order of the body, this is about as fast as it gets without major complication

Reply:
Is this mostly the same as `into_async_read()`? If so, can we use that instead of a loop?

Reply:
`.into_async_read()` doesn't allow for the same pre-allocation that we have, plus the loop's a pretty standard async pattern

Reply:
Got it. And we can't use `FuturesUnordered` because we do care about the order in this case. I'm still concerned that we're waiting in serial though - can we use [`FuturesOrdered`](https://docs.rs/futures/0.3.5/futures/stream/struct.FuturesOrdered.html) instead? That says

> While futures in the set will race to completion in parallel, results will only be returned in the order their originating futures were added to the queue.

> Futures are pushed into this queue and their realized values are yielded in order. This structure is optimized to manage a large number of futures. Futures managed by FuturesOrdered will only be polled when they generate notifications. This reduces the required amount of work needed to coordinate large numbers of futures.

Reply:
`FuturesUnordered` isn't used here, only for uploading batched files while this is downloading the contents of a singular future